task01/src/Vector.cpp

```diff
@@ -0,0 +1,318 @@
+    m_array = new int[m_capacity];
+    
+    // Initialize each element of the array with the value provided
+    for (int i = 0; i < m_size; i++)
+    {
+        m_array[i] = value;
+    }
+
+}
+
+//--------------------------------------------------------------------------------
+// Copy constructor
+Vector::Vector(const Vector& other)
+{
+    // Copy the size and capacity of the other vector to the lvalue vector's member variables
+    m_size = other.m_size;
+    m_capacity = other.m_capacity;
+
+    // Allocate a new array for the lvalue vector using the capacity of the other vector
+    m_array = new int[m_capacity];
+
+    // Copy the elements of the other vector's array to the lvalue vector's array
+    for (std::size_t i = 0; i < m_size; ++i) {
+        m_array[i] = other.m_array[i];
+    }
+
+}
+//--------------------------------------------------------------------------------
+/*copy assignment operator*/
+Vector& Vector::operator=(const Vector& other)
+{
+    // Check for self-assignment
+    if (this != &other)
+    {
+        // Resize the lvalue vector's array if necessary
+        if (m_size != other.m_size)
+        {
+            // Allocate a new array with the same size as the other vector's array
+            int* new_array = new int[other.m_size];
+
+            // Copy the elements of the other vector's array to the new array
+            for (std::size_t i = 0; i < other.m_size; ++i) {
+                new_array[i] = other.m_array[i];
+            }
+
+            // Deallocate the old array and update the lvalue vector's member variables
+            delete[] m_array;
+            m_array = new_array;
+            m_size = other.m_size;
+            m_capacity = other.m_size;
+        }
+        else // No need to resize the array
+        {
+            // Copy the elements of the other vector's array to the lvalue vector's array
+            for (std::size_t i = 0; i < m_size; ++i) {
+                m_array[i] = other.m_array[i];
+            }
+        }
+
+    }
+    return *this;
+}
+
+//--------------------------------------------------------------------------------
+/*move assignment operator, receives a temporary rvalue other and moves it to lvalue*/
+Vector& Vector::operator=(Vector&& other) noexcept {
+    /*std::cout << "move operator" << std::endl;*/
+    if (this != &other) {
+        delete[] m_array;
+        m_size = other.m_size;
+        m_capacity = other.m_capacity;
+        m_array = other.m_array;
+        other.m_size = 0;
+        other.m_capacity = 0;
+        other.m_array = nullptr;
+    }
+    return *this;
+}
+
+//--------------------------------------------------------------------------------
+
+void Vector::push_back(const int& value) {
+    // Check if we need to resize the array
+    if (m_size >= m_capacity) {
+        // Double the capacity to reduce the number of reallocations
+        std::size_t new_capacity = m_capacity * 2;
+        int* new_array = new int[new_capacity];
```

Comment:
Implementing this logic in `push_back()` creates code duplication with the `resize()` method. Why not use a common `reserve()` method that both will use and spare code duplication that way?